src/awkward/_v2/cpp-headers/awkward/LayoutBuilder.h

```diff
@@ -0,0 +1,1706 @@
+    }
+
+    void
+    set_parameters(std::string parameter) noexcept {
+      parameters_ = parameter;
+    }
+
+    bool
+    is_valid(std::string& error) const {
+      if (content_.length() != length()) {
+        std::stringstream out;
+        out << "BitMasked node" << id_ << "has content length " << content_.length()
+            << "but bit mask length " << mask_.length();
+        error.append(out.str());
+        return false;
+      }
+      else {
+        return content_.is_valid(error);
+      }
+    }
+
+    bool
+    valid_when() {
+      return valid_when_;
+    }
+
+    bool
+    lsb_order() {
+      return lsb_order_;
+    }
+
+    BUILDER&
+    append_valid() noexcept {
+      _append_begin();
+      current_byte_ |= cast_[current_index_];
+      _append_end();
+      return content_;
+    }
+
+    BUILDER&
+    extend_valid(size_t size) noexcept {
+      for (size_t i = 0; i < size; i++) {
+        append_valid();
+      }
+      return content_;
+    }
+
+    BUILDER&
+    append_null() noexcept {
+      _append_begin();
+      _append_end();
+      return content_;
+    }
+
+    BUILDER&
+    extend_null(size_t size) noexcept {
+      for (size_t i = 0; i < size; i++) {
+      append_null();
+      }
+      return content_;
+    }
+
+    void
+    buffer_nbytes(std::map<std::string, size_t> &names_nbytes) const noexcept {
+      names_nbytes["node" + std::to_string(id_) + "-mask"] = mask_.nbytes();
+      content_.buffer_nbytes(names_nbytes);
+    }
+
+    void
+    to_buffers(std::map<std::string, void*> &buffers) const noexcept {
+      mask_.concatenate(static_cast<uint8_t*>(buffers["node" + std::to_string(id_) + "-mask"]));
+      content_.to_buffers(buffers);
+    }
+
+    void
+    to_buffer(uint8_t* ptr) const noexcept {
+      mask_.concatenate(ptr);
+    }
+
+    std::string
+    form() const noexcept {
+      std::stringstream form_key, form_valid_when, form_lsb_order;
+      form_key << "node" << id_;
+      form_valid_when << std::boolalpha << valid_when_;
+      form_lsb_order << std::boolalpha << lsb_order_;
+      std::string params("");
+      if (parameters_ == "") { }
+      else {
+        params = std::string(", \"parameters\": " + parameters_);
+      }
+      return "{ \"class\": \"BitMaskedArray\", \"mask\": \"u8\", \"content\": " + content_.form()
+                + ", \"valid_when\": " + form_valid_when.str() + ", \"lsb_order\": "
+                + form_lsb_order.str() + params + ", \"form_key\": \"" + form_key.str() + "\" }";
+    }
+
+  private:
+    void
+    _append_begin() {
+      if (current_index_ == 8) {
+        current_byte_ = uint8_t(0);
+        current_byte_ref_ = mask_.append_and_get_ref(current_byte_);
+        current_index_ = 0;
+      }
+    }
+
+    void
+    _append_end() {
+      current_index_ += 1;
+      if (valid_when_) {
+        current_byte_ref_ = current_byte_;
+      }
+      else {
+        current_byte_ref_ = ~current_byte_;
+      }
+    }
```

Comment:
These functions didn't have to start with underscores because C++ is already marking them as private.

It also doesn't hurt, though.